consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/jsonrpc/methods/IbftGetSignerMetricsTest.java

```diff
@@ -0,0 +1,269 @@
+
+    when(blockchainQueries.headBlockNumber()).thenReturn(endBlock);
+
+    LongStream.range(startBlock, endBlock)
+        .forEach(value -> signerMetricResultList.add(generateBlock(value)));
+
+    final JsonRpcRequest request = requestWithParams(String.valueOf(startBlock), "latest");
+
+    final JsonRpcSuccessResponse response = (JsonRpcSuccessResponse) method.response(request);
+
+    LongStream.range(startBlock, endBlock)
+        .forEach(value -> verify(blockchainQueries).getBlockHeaderByNumber(value));
+
+    verify(blockchainQueries).headBlockNumber();
+
+    signerMetricResultList.forEach(
+        signerMetricResult ->
+            assertThat(response.getResult().toString()).contains(signerMetricResult.toString()));
+  }
+
+  @Test
+  public void getSignerMetricsWithPending() {
+
+    Long startBlock = 1L;
+    Long endBlock = 3L;
+
+    List<SignerMetricResult> signerMetricResultList = new java.util.ArrayList<>();
+
+    when(blockchainQueries.headBlockNumber()).thenReturn(endBlock);
+
+    LongStream.range(startBlock, endBlock)
+        .forEach(value -> signerMetricResultList.add(generateBlock(value)));
+
+    final JsonRpcRequest request = requestWithParams(String.valueOf(startBlock), "pending");
+
+    final JsonRpcSuccessResponse response = (JsonRpcSuccessResponse) method.response(request);
+
+    LongStream.range(startBlock, endBlock)
+        .forEach(value -> verify(blockchainQueries).getBlockHeaderByNumber(value));
+
+    verify(blockchainQueries).headBlockNumber();
+
+    signerMetricResultList.forEach(
+        signerMetricResult ->
+            assertThat(response.getResult().toString()).contains(signerMetricResult.toString()));
+  }
+
+  @Test
+  public void getSignerMetricsWithEarliest() {
+
+    Long startBlock = 0L;
+    Long endBlock = 3L;
+
+    List<SignerMetricResult> signerMetricResultList = new java.util.ArrayList<>();
+
+    LongStream.range(startBlock, endBlock)
+        .forEach(value -> signerMetricResultList.add(generateBlock(value)));
+
+    final JsonRpcRequest request = requestWithParams("earliest", String.valueOf(endBlock));
+
+    final JsonRpcSuccessResponse response = (JsonRpcSuccessResponse) method.response(request);
+
+    LongStream.range(startBlock, endBlock)
+        .forEach(value -> verify(blockchainQueries).getBlockHeaderByNumber(value));
+
+    verify(blockchainQueries, never()).headBlockNumber();
+
+    signerMetricResultList.forEach(
+        signerMetricResult ->
+            assertThat(response.getResult().toString()).contains(signerMetricResult.toString()));
+  }
+
+  private JsonRpcRequest requestWithParams(final Object... params) {
+    return new JsonRpcRequest(JSON_RPC_VERSION, IBFT_METHOD, params);
+  }
+
+  private SignerMetricResult generateBlock(final long number) {
+    final SECP256K1.KeyPair proposerKeyPair = SECP256K1.KeyPair.generate();
+    final Address proposerAddressBlock = Util.publicKeyToAddress(proposerKeyPair.getPublicKey());
```

Comment:
Creating keypairs is fairly expensive and we really want to test that the number of blocks gets incremented so it's probably worth creating a few key pairs as static fields which are just iterated through here (e.g. `final SECP256K1.KeyPair proposerKeyPair = KEYPAIRS[number % KEYPAIRS.length]`)

Reply:
done